madminer/utils/morphing.py

```diff
@@ -123,41 +130,127 @@ def find_components(self, max_overall_power=4):
             The maximal sum of powers of all parameters contributing to the squared matrix element.
             Typically, if parameters can affect the couplings at n vertices, this number is 2n. Default value: 4.
 
+        BSM_max_power : int, optional
+            The maximal sum of powers of all parameters contributing to the squared matrix element for the BSM couplings.
+
+        Np : int, optional
+            The number of parameters affecting the production vertex.
+
+        Nd : int, optional
+            The number of parameters affecting the decay vertex.
+
+        Ns : int, optional
+            The number of parameters that affect both the production and decay vertices.
+
         Returns
         -------
         components : ndarray
             Array with shape (n_components, n_parameters), where each entry gives the power with which a parameter
             scales a given component.
         """
+        if Nd != 0 or Np != 0 or Ns != 0:
+            components = []
+            # Check if any number of couplings were specified.
+            if Nd == 0 and Np == 0 and Ns == 0:
+                raise RuntimeError("Coupling numbers not specified")
+
+            # number of couplings
+            gp = sp.symbols("gp:15")
+            gd = sp.symbols("gd:15")
+            gs = sp.symbols("gs:15")
+
+            if Ns != 0:
+                prod = sum(gp[:Np] + gs[:Ns])  # sum of couplings in production
+                dec = sum(gd[:Nd] + gs[:Ns])  # sum of couplings in decay
+            else:
+                prod = sum(gp[:Np])  # sum of couplings in production
+                dec = sum(gd[:Nd])  # sum of couplings in decay
+
+            if Nd == 0 and Ns == 0:
+                dec = 1
+            if Np == 0 and Ns == 0:
+                prod = 1
+            f = sp.expand((prod) ** 2 * (dec) ** 2)  # contribution to matrix element squared
+            mono = sp.Poly(f).terms()
+            # list of tuples containing monomials
+
+            for i in range(0, len(mono)):
+                components.append(mono[i][0])
```

Comment:
Three comments here:

1. The iterator can be simplified: `range(0, len(...))` --> `range(len(...))` --> `enumerate(...)`
2. If you are iterating on an array just to access its elements (i.e. `mono[i]`), iterate through them directly
3. It seems the `components` array is not used, anywhere, before this point. Why not creating it here?

```suggestion
            components = [mono_value[0] for mono_value in mono]
```

Reply:
Thank you for the feedbacks! I've made the changes and also move the components down here.